ci: expand build matrix

bcressey opened the pull request:
**Issue number:**
N/A


**Description of changes:**
Adds coverage for aarch64 and both Kubernetes variants.


**Testing done:**
Tested via PR.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
ci: expand build matrix

Adds coverage for aarch64 and both Kubernetes variants.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

tjkirch reviewed (commented):
It looks like the right change, but the question is whether we want to merge it before fixing the build issue causing flaky failures unrelated to our changes; we'd be four times as likely to show a useless error...

bcressey commented:
> It looks like the right change, but the question is whether we want to merge it before fixing the build issue causing flaky failures unrelated to our changes; we'd be four times as likely to show a useless error...

+1

This also doubling as a stress test to help me chase down the problem. I had a faint hope that the AMI refresh for the builders would help, but no luck.

Commit:
build: drop unused BUILDSYS_DOCKER_RUN_ARGS variable

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
build: add BUILDSYS_JOBS variable

Limiting ourselves to eight concurrent package builds has worked OK
in practice, even though there's no scientific basis to the number.

Making it a variable allows us to increase parallelism in scenarios
where we control the build fleet and can measure the results.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

tjkirch reviewed (approved):
If controlling the job count helps with our failures, as it seems to, I'm happy!

jahkeup commented on .github/workflows/build.yml:
```diff
@@ -5,8 +5,13 @@ on:
     runs-on: [self-hosted, linux, x64]
+    strategy:
+      matrix:
+        variant: [aws-k8s-1.15, aws-k8s-1.16]
```
Any reason, other than total run time, to not include `aws-dev` in this list?

bcressey replied:
I'm somewhat reluctant to promote it as a "Tier 1" target which is how I view this list. All of the packages it needs still get built, so we'd catch docker build failures for example.

jahkeup replied:
Gotcha. I think we should still be building the variant as much as possible - perhaps we could include it in a non-blocking/non-critical way where it isn't required for PR to pass?

jahkeup commented on Makefile.toml:
```diff
@@ -12,28 +12,33 @@ BUILDSYS_TIMESTAMP = { script = ["date +%s"] }
 BUILDSYS_ALLOW_FAILED_LICENSE_CHECK = "false"
 
+# This controls how many `docker build` commands we'll invoke at once.
+BUILDSYS_JOBS = "8"
```
Is this number respected in downstream builds (eg: `make -j1` for each respective process)?

bcressey replied:
It's not - they are somewhat orthogonal concepts. The kernel build can still make use of all cores and we want to keep that property even if we only have a few builds going at once.

jahkeup replied:
I see. I read this as though the parameter were prefixed with `BUILDSYS_` where `BUILDSYS_JOBS` meant `JOBS` (as in `-j`) which I'm used to seeing respected up and down the build "call stack".

jahkeup commented:
Darn - I clicked re-run wanting only the failed `aws-k8s-1.16, aarch64` build to be re-attempted. Sorry about the pending builds!

Commit:
buildsys: retry builds on frontend error

Work around a transient failure that makes CI builds unreliable.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

tjkirch commented on tools/buildsys/src/builder.rs:
```diff
@@ -140,7 +159,21 @@ fn build(target: &str, build_args: &str, tag: &str, output: &str) -> Result<()>
 
     // Build the image, which builds the artifacts we want.
-    docker(&build)?;
+    // HACK: work around a transient, known failure case with Docker.
```
Is it worth adding this capability to the `docker` function?  I could imagine a signature like:

```
fn docker(args: &[String], retry: Retry) {...}

enum Retry {
   No,
   Yes {
      count: NonZeroU8,
      retry_failure_messages: &[&str],
   }
}
```

bcressey replied:
That would be neat, not sure if it's worth it right now. Maybe when we have more than one known error?

tjkirch replied:
I think my concern was more around the repeated code and the need for a "hack" than the list of strings specifically, but obviously I don't think it's a blocker.  It'll be nice to get around this issue.
